docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,251 @@
+
+Firstly, the query results cache needs to be invalidated for each new delete request or a cancellation of one. This can be accomplished by utilizing cache generation numbers. For each tenant, their cache is prefixed with a cache generation number. When the query front-end discovers a cache generation number that is greater than the previous generation number, then it knows to invalidate the query results cache. However, the cache can only be invalidated once the queriers have loaded the tombstones from the bucket index and have begun filtering the data. Otherwise, to-be deleted data might show up in queries and be cached again. One of the way to guarantee that all the queriers are using the new tombstones is to wait until the bucket index staleness period has passed from the time the tombstones have been written to the bucket index. The staleness period can be configured using the following flag: `-blocks-storage.bucket-store.bucket-index.max-stale-period`. We can use the bucket index staleness period as the delay to wait before the cache generation number is increased. A query will fail inside the querier, if the bucket index last update is older the staleness period. Once this period is over, all the queriers should have the updated tombstones and the query results cache can be invalidated.  Here is the proposed method for accomplishing this:
+
+
+- The cache generation number will be a timestamp. It will also serve as the time of when it becomes valid and the query front-end can use it.
+- The bucket index will store the cache generation number.  The query front-end will periodically fetch the bucket index.
+- Inside the compactor, it will load the tombstones from object store and update the bucket index accordingly. If a deletion request is made or cancelled, the compactor will discover this and increment the cache generation number in the bucket index. The cache generation number will be the current timestamp + the max stale period of the bucket index.  The compactor can discover if there have been any changes to the tombstones by comparing the newly loaded tombstones to the one's currently in the bucket index.
+- The query front-end will fetch the cache generation number from the bucket index. If the current timestamp is less than the cache generation number, it will simply not do anything as the generation number is not yet valid. If the query front-end discovers that the current time has passed the cache generation timestamp from the bucket index, then it is valid and can be used. The query front end will compare it to the current cache generation number stored in the front-end. If the cache generation number from the front-end is less than the one from bucket index, then the cache is invalidated.
```

Comment:
I'm not sure this works. Think about the case the query-frontend is restarted. At startup it loads the bucket index for a tenant to get the cache generation number and that number is in the future. What to do? It can't use the cache generation number in the bucket index, but it doesn't even know which was the previous one.

The proposed approach also has another problem: if the max stale period is 1h and you create a tombstone every < 1h, then the cache is never invalidated.

I have a different proposal:
- The cache generation number in the bucket index is the current one (the one the query-frontend has to use)
- If the bucket index has no cache generation number (think about an old version of the bucket index, when you rollout this feature) than the generation number is a default `0`
- The compactor, during the blocks cleanup process, looks all tombstones and compute the current generation number being the max timestamp across all tombstones created until "now - max stale period"
- What's about cancellations? We need to invalidate the cache also after cancellations. We could change the process of how a tombstone is deleted. We don't immediately delete the tombstone from the object storage, but we switch its state to `.deleted` (adding a `deletionTime` too) . This allows the compactor to also be aware of deleted tombstones and when it was done. Finally, the compactor's blocks cleaner can also take care of cleaning up `.deleted` tombstones after X period of time (eg. 10 times max stale period).

Reply:
That makes a lot of sense, thanks for the suggestions! Proposal is updated now. Only difference is instead of using a `deletionTime`, I added another field called `stateCreationTime` which can serve in the same way. This way it can also be used for seeing how much time elapsed since moving into the pending state.